In [1]:
#import libraries
import pandas as pd
import numpy as np

In [2]:
#Read the movie Dataset
i_cols = ['movie id', 'movie title' ,'release date','video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure',
    'Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy',
    'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
items = pd.read_csv('ml-100k/u.item', sep='|', names=i_cols,encoding='latin-1')

In [3]:
items

movie id                                movie title release date  \
0            1                           Toy Story (1995)  01-Jan-1995   
1            2                           GoldenEye (1995)  01-Jan-1995   
2            3                          Four Rooms (1995)  01-Jan-1995   
3            4                          Get Shorty (1995)  01-Jan-1995   
4            5                             Copycat (1995)  01-Jan-1995   
...        ...                                        ...          ...   
1677      1678                          Mat' i syn (1997)  06-Feb-1998   
1678      1679                           B. Monkey (1998)  06-Feb-1998   
1679      1680                       Sliding Doors (1998)  01-Jan-1998   
1680      1681                        You So Crazy (1994)  01-Jan-1994   
1681      1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

      video release date                                           IMDb URL  \
0                    NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...   
1                    NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...   
2                    NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...   
3                    NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...   
4                    NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)   
...                  ...                                                ...   
1677                 NaN  http://us.imdb.com/M/title-exact?Mat%27+i+syn+...   
1678                 NaN  http://us.imdb.com/M/title-exact?B%2E+Monkey+(...   
1679                 NaN      http://us.imdb.com/Title?Sliding+Doors+(1998)   
1680                 NaN  http://us.imdb.com/M/title-exact?You%20So%20Cr...   
1681                 NaN  http://us.imdb.com/M/title-exact?Schrei%20aus%...   

      unknown  Action  Adventure  Animation  Children's  ...  Fantasy  \
0           0       0          0          1           1  ...        0   
1           0       1          1          0           0  ...        0   
2           0       0          0          0           0  ...        0   
3           0       1          0          0           0  ...        0   
4           0       0          0          0           0  ...        0   
...       ...     ...        ...        ...         ...  ...      ...   
1677        0       0          0          0           0  ...        0   
1678        0       0          0          0           0  ...        0   
1679        0       0          0          0           0  ...        0   
1680        0       0          0          0           0  ...        0   
1681        0       0          0          0           0  ...        0   

      Film-Noir  Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  \
0             0       0        0        0        0       0         0    0   
1             0       0        0        0        0       0         1    0   
2             0       0        0        0        0       0         1    0   
3             0       0        0        0        0       0         0    0   
4             0       0        0        0        0       0         1    0   
...         ...     ...      ...      ...      ...     ...       ...  ...   
1677          0       0        0        0        0       0         0    0   
1678          0       0        0        0        1       0         1    0   
1679          0       0        0        0        1       0         0    0   
1680          0       0        0        0        0       0         0    0   
1681          0       0        0        0        0       0         0    0   

      Western  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
1677        0  
1678        0  
1679        0  
1680        0  
1681        0  

[1682 rows x 24 columns]

In [4]:
# Select only needed Item
movie_content=items[['movie id', 'movie title' , 'Action', 'Adventure','Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy','Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']]

In [5]:
movie_content

movie id                                movie title  Action  Adventure  \
0            1                           Toy Story (1995)       0          0   
1            2                           GoldenEye (1995)       1          1   
2            3                          Four Rooms (1995)       0          0   
3            4                          Get Shorty (1995)       1          0   
4            5                             Copycat (1995)       0          0   
...        ...                                        ...     ...        ...   
1677      1678                          Mat' i syn (1997)       0          0   
1678      1679                           B. Monkey (1998)       0          0   
1679      1680                       Sliding Doors (1998)       0          0   
1680      1681                        You So Crazy (1994)       0          0   
1681      1682  Scream of Stone (Schrei aus Stein) (1991)       0          0   

      Animation  Children's  Comedy  Crime  Documentary  Drama  Fantasy  \
0             1           1       1      0            0      0        0   
1             0           0       0      0            0      0        0   
2             0           0       0      0            0      0        0   
3             0           0       1      0            0      1        0   
4             0           0       0      1            0      1        0   
...         ...         ...     ...    ...          ...    ...      ...   
1677          0           0       0      0            0      1        0   
1678          0           0       0      0            0      0        0   
1679          0           0       0      0            0      1        0   
1680          0           0       1      0            0      0        0   
1681          0           0       0      0            0      1        0   

      Film-Noir  Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  \
0             0       0        0        0        0       0         0    0   
1             0       0        0        0        0       0         1    0   
2             0       0        0        0        0       0         1    0   
3             0       0        0        0        0       0         0    0   
4             0       0        0        0        0       0         1    0   
...         ...     ...      ...      ...      ...     ...       ...  ...   
1677          0       0        0        0        0       0         0    0   
1678          0       0        0        0        1       0         1    0   
1679          0       0        0        0        1       0         0    0   
1680          0       0        0        0        0       0         0    0   
1681          0       0        0        0        0       0         0    0   

      Western  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
1677        0  
1678        0  
1679        0  
1680        0  
1681        0  

[1682 rows x 20 columns]

In [6]:
#read the ratings dataset
r_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv('ml-100k/u.data', sep='\t', names=r_cols,encoding='latin-1')

In [47]:
ratings

user_id  movie_id  rating  unix_timestamp
0          196       242       3       881250949
1          186       302       3       891717742
2           22       377       1       878887116
3          244        51       2       880606923
4          166       346       1       886397596
...        ...       ...     ...             ...
99995      880       476       3       880175444
99996      716       204       5       879795543
99997      276      1090       1       874795795
99998       13       225       2       882399156
99999       12       203       3       879959583

[100000 rows x 4 columns]

In [91]:
ratings['movie_id'].value_counts()

movie_id
50      583
258     509
100     508
181     507
294     485
       ... 
1647      1
1618      1
1632      1
1627      1
1636      1
Name: count, Length: 1682, dtype: int64

# Here we have 1682 unique movies, which means same movies are being rated by different users, inorder to consolidate the multiple ratings given for a single movie we find the average rating 

In [49]:
avg_ratings=ratings.groupby('movie_id')['rating'].mean().reset_index()
avg_ratings

movie_id    rating
0            1  3.878319
1            2  3.206107
2            3  3.033333
3            4  3.550239
4            5  3.302326
...        ...       ...
1677      1678  1.000000
1678      1679  3.000000
1679      1680  2.000000
1680      1681  3.000000
1681      1682  3.000000

[1682 rows x 2 columns]

In [70]:
avg_ratings.rename(columns={'rating':'avg_ratings','movie_id':'movie id'},inplace=True)

In [71]:
avg_ratings

movie id  avg_ratings
0            1     3.878319
1            2     3.206107
2            3     3.033333
3            4     3.550239
4            5     3.302326
...        ...          ...
1677      1678     1.000000
1678      1679     3.000000
1679      1680     2.000000
1680      1681     3.000000
1681      1682     3.000000

[1682 rows x 2 columns]

In [72]:
top_rated=avg_ratings[avg_ratings['avg_ratings']>4]
top_rated

movie id  avg_ratings
11          12     4.385768
21          22     4.151515
22          23     4.120879
44          45     4.050000
47          48     4.094017
...        ...          ...
1593      1594     4.500000
1598      1599     5.000000
1638      1639     4.333333
1641      1642     4.500000
1652      1653     5.000000

[132 rows x 2 columns]

# Now i need to compare these movie id in list with movie set to get the category of movies

In [73]:
# Merging both movie and top rated dataframe
merge= high_rated_movies = pd.merge(top_rated, movie_content,on='movie id')

In [74]:
merge

movie id  avg_ratings                                        movie title  \
0          12     4.385768                         Usual Suspects, The (1995)   
1          22     4.151515                                  Braveheart (1995)   
2          23     4.120879                                 Taxi Driver (1976)   
3          45     4.050000                         Eat Drink Man Woman (1994)   
4          48     4.094017                                 Hoop Dreams (1994)   
..        ...          ...                                                ...   
127      1594     4.500000                                     Everest (1998)   
128      1599     5.000000                      Someone Else's America (1995)   
129      1639     4.333333                Bitter Sugar (Azucar Amargo) (1996)   
130      1642     4.500000                           Some Mother's Son (1996)   
131      1653     5.000000  Entertaining Angels: The Dorothy Day Story (1996)   

     Action  Adventure  Animation  Children's  Comedy  Crime  Documentary  \
0         0          0          0           0       0      1            0   
1         1          0          0           0       0      0            0   
2         0          0          0           0       0      0            0   
3         0          0          0           0       1      0            0   
4         0          0          0           0       0      0            1   
..      ...        ...        ...         ...     ...    ...          ...   
127       0          0          0           0       0      0            1   
128       0          0          0           0       0      0            0   
129       0          0          0           0       0      0            0   
130       0          0          0           0       0      0            0   
131       0          0          0           0       0      0            0   

     ...  Fantasy  Film-Noir  Horror  Musical  Mystery  Romance  Sci-Fi  \
0    ...        0          0       0        0        0        0       0   
1    ...        0          0       0        0        0        0       0   
2    ...        0          0       0        0        0        0       0   
3    ...        0          0       0        0        0        0       0   
4    ...        0          0       0        0        0        0       0   
..   ...      ...        ...     ...      ...      ...      ...     ...   
127  ...        0          0       0        0        0        0       0   
128  ...        0          0       0        0        0        0       0   
129  ...        0          0       0        0        0        0       0   
130  ...        0          0       0        0        0        0       0   
131  ...        0          0       0        0        0        0       0   

     Thriller  War  Western  
0           1    0        0  
1           0    1        0  
2           1    0        0  
3           0    0        0  
4           0    0        0  
..        ...  ...      ...  
127         0    0        0  
128         0    0        0  
129         0    0        0  
130         0    0        0  
131         0    0        0  

[132 rows x 21 columns]

In [133]:
#Check for single category
merge[merge['Action']==1].sort_values(by='avg_ratings',ascending=False).head(5)['movie title'].values
pd.DataFrame(merge[merge['Action']==1].sort_values(by='avg_ratings',ascending=False).head(5)['movie title'].values, columns=['movie title'])

movie title
0                 Star Wars (1977)
1            Godfather, The (1972)
2   Raiders of the Lost Ark (1981)
3                   Titanic (1997)
4  Empire Strikes Back, The (1980)

In [90]:
list=['Action', 'Adventure', 'Animation', "Children's", 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
for columns in list:
    content=merge[merge[columns]==1].sort_values(by='avg_ratings',ascending=False).head(5)['movie title']
    print('\n',columns)
    print('\n',content)


 Action

 5                    Star Wars (1977)
20              Godfather, The (1972)
31     Raiders of the Lost Ark (1981)
52                     Titanic (1997)
29    Empire Strikes Back, The (1980)
Name: movie title, dtype: object

 Adventure

 118                    Star Kid (1997)
5                     Star Wars (1977)
31      Raiders of the Lost Ark (1981)
73           Lawrence of Arabia (1962)
29     Empire Strikes Back, The (1980)
Name: movie title, dtype: object

 Animation

 58                                 Close Shave, A (1995)
27                            Wrong Trousers, The (1993)
17     Wallace & Gromit: The Best of Aardman Animatio...
119                                         Faust (1994)
38                               Grand Day Out, A (1992)
Name: movie title, dtype: object

 Children's

 118             Star Kid (1997)
21     Wizard of Oz, The (1939)
Name: movie title, dtype: object

 Comedy

 124     Santa with Muscles (1996)
58          Close Shave, A (1995)
2

In [125]:
list=[col for col in movie_content.columns if col not in ['movie id', 'movie title']]
#list=['Action', 'Adventure', 'Animation', "Children's", 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
for columns in list:
    content=merge[merge[columns]==1].sort_values(by='avg_ratings',ascending=False).head(5)['movie title']
    print('\n',"TOP 5",columns,"MOVIES")
    print('\n',content)


 TOP 5 Action MOVIES

 5                    Star Wars (1977)
20              Godfather, The (1972)
31     Raiders of the Lost Ark (1981)
52                     Titanic (1997)
29    Empire Strikes Back, The (1980)
Name: movie title, dtype: object

 TOP 5 Adventure MOVIES

 118                    Star Kid (1997)
5                     Star Wars (1977)
31      Raiders of the Lost Ark (1981)
73           Lawrence of Arabia (1962)
29     Empire Strikes Back, The (1980)
Name: movie title, dtype: object

 TOP 5 Animation MOVIES

 58                                 Close Shave, A (1995)
27                            Wrong Trousers, The (1993)
17     Wallace & Gromit: The Best of Aardman Animatio...
119                                         Faust (1994)
38                               Grand Day Out, A (1992)
Name: movie title, dtype: object

 TOP 5 Children's MOVIES

 118             Star Kid (1997)
21     Wizard of Oz, The (1939)
Name: movie title, dtype: object

 TOP 5 Comedy MOVIES

 124 

In [130]:
def popularity_based(ratings,movie_content,threshold):
    avg_ratings=ratings.groupby('movie_id')['rating'].mean().reset_index()
    avg_ratings.rename(columns={'rating':'avg_ratings','movie_id':'movie id'},inplace=True)
    #avg_ratings
    top_rated=avg_ratings[avg_ratings['avg_ratings']>threshold]
    #top_rated
    # Merging both movie and top rated dataframe
    merge= high_rated_movies = pd.merge(top_rated, movie_content,on='movie id')
    list=[col for col in movie_content.columns if col not in ['movie id', 'movie title']]
    #list=['Action', 'Adventure', 'Animation', "Children's", 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
    for columns in list:
        content=merge[merge[columns]==1].sort_values(by='avg_ratings',ascending=False).head(5)['movie title'].values
        content_table=pd.DataFrame(content,columns=['movie title'])
        print('\n',"TOP 5",columns,"MOVIES:")
        print('\n',content_table)

 

In [132]:
popularity_based(ratings,movie_content,3.5)


 TOP 5 Action MOVIES:

                        movie title
0                 Star Wars (1977)
1            Godfather, The (1972)
2   Raiders of the Lost Ark (1981)
3                   Titanic (1997)
4  Empire Strikes Back, The (1980)

 TOP 5 Adventure MOVIES:

                        movie title
0                  Star Kid (1997)
1                 Star Wars (1977)
2   Raiders of the Lost Ark (1981)
3        Lawrence of Arabia (1962)
4  Empire Strikes Back, The (1980)

 TOP 5 Animation MOVIES:

                                          movie title
0                              Close Shave, A (1995)
1                         Wrong Trousers, The (1993)
2  Wallace & Gromit: The Best of Aardman Animatio...
3                                       Faust (1994)
4                            Grand Day Out, A (1992)

 TOP 5 Children's MOVIES:

                 movie title
0           Star Kid (1997)
1  Wizard of Oz, The (1939)
2               Babe (1995)
3      Fly Away Home (1996)
4          T